staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -264,19 +460,217 @@ func (q *delayingType) waitingLoop() {
 	}
 }
 
-// insert adds the entry to the priority queue, or updates the readyAt if it already exists in the queue
-func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
-	// if the entry already exists, update the time only if it would cause the item to be queued sooner
+// insert adds the entry to the 'waiting' queue, or updates the readyAt if it already exists in the queue
+// It returns a boolean to indicate that the entry was inserted or not.
+func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) bool {
+	// if the entry already exists, update the time according to the entry.options.WhenWaiting
 	existing, exists := knownEntries[entry.data]
 	if exists {
-		if existing.readyAt.After(entry.readyAt) {
+		var replace bool
+		switch entry.options.WhenWaiting {
+		case TakeShorter:
+			replace = existing.readyAt.After(entry.readyAt)
+		case TakeLonger:
+			replace = existing.readyAt.Before(entry.readyAt)
+		case TakeIncoming:
+			replace = !existing.readyAt.Equal(entry.readyAt)
+		}
+		if replace {
 			existing.readyAt = entry.readyAt
+			existing.options = entry.options
 			heap.Fix(q, existing.index)
+			return true
 		}
-
-		return
+		return false
 	}
 
 	heap.Push(q, entry)
 	knownEntries[entry.data] = entry
+	return true
+}
+
+// calculateNextReadyAt recalculates the nextReadyAtTime as follows:
+// - removes the timer when the 'waiting' queue is empty.
+// - resets the timer when the head 'waiting' item has changed.
+// - creates a new timer when 1+ items are queued to an empty 'waiting' queue.
+func (q *delayingType) calculateNextReadyAt(w *waitForPriorityQueue) {
+	var head *waitFor
+
+	// early escape when the timer has not changed
+	if w.Len() > 0 {
+		head = w.Peek().(*waitFor)
+		if q.headReadyTime.Equal(head.readyAt) {
+			return
+		}
+	}
+
+	// reset timer
+	if q.nextReadyAtTimer != nil {
+		q.nextReadyAtTimerCh = make(<-chan time.Time)
+		q.nextReadyAtTimer.Stop()
+		q.nextReadyAtTimer = nil
+	}
+
+	// create new timer
+	if head != nil {
+		now := q.clock.Now()
+		q.nextReadyAtTimer = q.clock.NewTimer(head.readyAt.Sub(now))
+		q.nextReadyAtTimerCh = q.nextReadyAtTimer.C()
+		q.headReadyTime = head.readyAt
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: calculated NEXT-READY-AT timer " + head.readyAt.String()))
+		}
+	}
+}
+
+// closeAllWaitForFills loops through all the channels in waitSync and returns the time of the next head item
+// before closing each channel.
+func (q *delayingType) closeAllWaitForFills() {
+	if q.waitSync == nil {
+		return
+	}
+
+	// always report back the duration until the next ready item
+	r := waitingLoopResponse{}
+	if q.nextReadyAtTimer != nil {
```

Comment:
why the condition here? Don't we want to report in all cases? If not we can just test with `q.headReadyTime.IsZero()` instead

Reply:
Hmm, yes this condition should be the length of the queue - have refactored and will use len instead.